File: teaching/anthropic/anthropic-protocol-book/code/node/ch05/SSEParser.ts
/**
 * Chapter 5: SSE (Server-Sent Events) Parser (TypeScript).
 *
 * A dependency-free, incremental SSE parser that conforms to the WHATWG
 * Server-Sent Events specification. Handles chunked input, multi-line
 * data fields, keepalive pings, and boundary-spanning events.
 *
 * @example
 * ```ts
 * const parser = new SSEParser();
 * for (const event of parser.feed(chunk)) {
 *   console.log(event.event, event.data);
 * }
 * ```
 */

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** A single decoded SSE event ready for application consumption. */
export interface ParsedEvent {
  /** Event type (e.g. "message_start", "content_block_delta"). */
  readonly event: string;
  /** Parsed JSON data carried by the event. */
  readonly data: Record<string, unknown>;
}

/** Raised when an SSE event cannot be parsed correctly. */
export class SSEParseError extends Error {
  readonly rawData: string | null;

  constructor(message: string, rawData?: string) {
    super(message);
    this.name = "SSEParseError";
    this.rawData = rawData ?? null;
  }
}

// ---------------------------------------------------------------------------
// SSE Parser
// ---------------------------------------------------------------------------

/**
 * Incremental SSE (Server-Sent Events) parser.
 *
 * Feed raw data as it arrives from the network. The parser buffers
 * incomplete events internally and yields complete, parsed events
 * as they are detected.
 */
export class SSEParser {
  private buffer = "";

  /**
   * Feed a chunk of raw SSE data into the parser.
   *
   * @returns Zero or more complete `ParsedEvent` objects decoded from
   *          the stream so far.
   */
  feed(chunk: string): ParsedEvent[] {
    this.buffer += chunk.replace(/\r\n/g, "\n").replace(/\r/g, "\n");

    const events: ParsedEvent[] = [];

    // eslint-disable-next-line no-constant-condition
    while (true) {
      const idx = this.buffer.indexOf("\n\n");
      if (idx === -1) break;

      const eventText = this.buffer.slice(0, idx);
      this.buffer = this.buffer.slice(idx + 2); // skip the \n\n
      const parsed = this.parseEvent(eventText);
      if (parsed !== null) {
        events.push(parsed);
      }
    }

    return events;
  }

  /**
   * Return any remaining complete events in the buffer.
   *
   * Call this when the underlying stream has closed to drain the
   * last event (which may not end with `\n\n`).
   */
  flush(): ParsedEvent[] {
    if (this.buffer.trim() === "") {
      this.buffer = "";
      return [];
    }

    const events: ParsedEvent[] = [];
    const parsed = this.parseEvent(this.buffer);
    if (parsed !== null) {
      events.push(parsed);
    }
    this.buffer = "";
    return events;
  }

  /** Reset the parser to a pristine state (discards buffer). */
  reset(): void {
    this.buffer = "";
  }

  // ------------------------------------------------------------------
  // Internal helpers
  // ------------------------------------------------------------------

  /**
   * Parse a single SSE event from its text representation.
   *
   * Implements the WHATWG SSE parsing algorithm for a single
   * event block (text between two `\n\n` delimiters).
   *
   * @returns A `ParsedEvent`, or `null` for events that should be
   *          silently ignored (comments, keepalive pings).
   */
  private parseEvent(eventText: string): ParsedEvent | null {
    let eventType = "message";
    const dataParts: string[] = [];

    for (const line of eventText.split("\n")) {
      const trimmed = line.endsWith("\r") ? line.slice(0, -1) : line;

      // Empty line -- ignored
      if (trimmed === "") continue;

      // Comment line (starts with colon) -- ignored
      if (trimmed.startsWith(":")) continue;

      // Look for the colon separator
      const colonPos = trimmed.indexOf(":");
      let fieldName: string;
      let fieldValue: string;

      if (colonPos === -1) {
        // Field with no value
        fieldName = trimmed;
        fieldValue = "";
      } else {
        fieldName = trimmed.slice(0, colonPos);
        // Skip exactly one optional space after colon
        if (trimmed.length > colonPos + 1 && trimmed[colonPos + 1] === " ") {
          fieldValue = trimmed.slice(colonPos + 2);
        } else {
          fieldValue = trimmed.slice(colonPos + 1);
        }
      }

      if (fieldName === "event") {
        eventType = fieldValue;
      } else if (fieldName === "data") {
        dataParts.push(fieldValue);
      }
      // id and retry fields are accepted but not used by Anthropic's API
    }

    // If no data parts, treat as keepalive/comment
    if (dataParts.length === 0) return null;

    const mergedData = dataParts.join("\n");

    // Parse JSON
    let parsedData: Record<string, unknown>;
    try {
      parsedData = JSON.parse(mergedData) as Record<string, unknown>;
    } catch (err) {
      throw new SSEParseError(
        `Invalid JSON in SSE data field: ${err instanceof Error ? err.message : String(err)}`,
        mergedData,
      );
    }

    // Silently drop ping events -- transport-layer keepalives
    if (eventType === "ping" || parsedData["type"] === "ping") {
      return null;
    }

    return { event: eventType, data: parsedData };
  }
}

File: teaching/anthropic/anthropic-protocol-book/code/python/ch05/sse_parser.py
"""
Chapter 5: SSE (Server-Sent Events) Parser.

A dependency-free, incremental SSE parser that conforms to the WHATWG
Server-Sent Events specification.  Handles chunked input, multi-line
data fields, keepalive pings, and boundary-spanning events.

Usage::

    parser = SSEParser()
    for chunk in http_response.iter_bytes():
        for event in parser.feed(chunk):
            handle(event)
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any


# ---------------------------------------------------------------------------
# Public types
# ---------------------------------------------------------------------------


@dataclass
class ParsedEvent:
    """A single decoded SSE event ready for application consumption."""

    event: str
    """Event type (e.g. ``message_start``, ``content_block_delta``)."""

    data: dict[str, Any]
    """Parsed JSON data carried by the event."""


@dataclass
class SSEParseError(Exception):
    """Raised when an SSE event cannot be parsed correctly."""

    message: str
    raw_data: str | None = None


# ---------------------------------------------------------------------------
# SSE Parser
# ---------------------------------------------------------------------------


class SSEParser:
    """Incremental SSE (Server-Sent Events) parser.

    Feed raw bytes as they arrive from the network.  The parser
    buffers incomplete events internally and yields complete,
    parsed events as they are detected.

    Design decisions
    ----------------
    - **Zero allocations for idle connections:** ``feed(b\"\")`` is a
      near-no-op -- only a bounds check on the buffer.
    - **No regex.** The parser uses `.split()` and `.partition()` for speed.
    - **Comment/keepalive filtering.** ``ping`` events and comment lines
      are silently dropped -- they never appear in the output.
    """

    def __init__(self) -> None:
        self._buffer: str = ""

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def feed(self, chunk: bytes) -> list[ParsedEvent]:
        """Feed raw bytes into the parser.

        Returns a list of zero or more complete :class:`ParsedEvent`
        objects decoded from the stream so far.
        """
        # Normalize line endings: \r\n → \n, standalone \r → \n
        text = chunk.decode("utf-8", errors="replace")
        text = text.replace("\r\n", "\n").replace("\r", "\n")
        self._buffer += text

        events: list[ParsedEvent] = []
        while True:
            idx = self._buffer.find("\n\n")
            if idx == -1:
                break
            event_text = self._buffer[:idx]
            self._buffer = self._buffer[idx + 2 :]  # skip the \n\n
            parsed = self._parse_event(event_text)
            if parsed is not None:
                events.append(parsed)
        return events

    def flush(self) -> list[ParsedEvent]:
        """Return any remaining complete events in the buffer.

        Call this when the underlying stream has closed to drain the
        last event (which might not end with ``\\n\\n``).
        """
        if not self._buffer.strip():
            self._buffer = ""
            return []
        events: list[ParsedEvent] = []
        parsed = self._parse_event(self._buffer)
        if parsed is not None:
            events.append(parsed)
        self._buffer = ""
        return events

    def reset(self) -> None:
        """Reset the parser to a pristine state (discards buffer)."""
        self._buffer = ""

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _parse_event(event_text: str) -> ParsedEvent | None:
        """Parse a single SSE event from its text representation.

        Implements the WHATWG SSE parsing algorithm for a single
        event block (text between two ``\\n\\n`` delimiters).

        Returns ``None`` for events that should be silently ignored
        (comments, keepalive pings with no data).
        """
        event_type: str = "message"
        data_parts: list[str] = []

        for line in event_text.split("\n"):
            line = line.rstrip("\r")

            # Empty line -- ignored
            if line == "":
                continue

            # Comment line (starts with colon) -- ignored
            if line.startswith(":"):
                continue

            # Look for the colon separator
            colon_pos = line.find(":")
            if colon_pos == -1:
                # Field with no value -- treat as empty field
                field_name = line
                field_value = ""
            else:
                field_name = line[:colon_pos]
                # Skip exactly one optional space after colon
                if len(line) > colon_pos + 1 and line[colon_pos + 1] == " ":
                    field_value = line[colon_pos + 2 :]
                else:
                    field_value = line[colon_pos + 1 :]

            if field_name == "event":
                event_type = field_value
            elif field_name == "data":
                data_parts.append(field_value)
            # id and retry fields are accepted but not used by
            # Anthropic's API, so we ignore them.

        # If no data parts, treat as keepalive/comment
        if not data_parts:
            return None

        # Merge multiple data lines with LF
        merged_data = "\n".join(data_parts)

        # Parse JSON
        try:
            parsed_data: dict[str, Any] = json.loads(merged_data)
        except json.JSONDecodeError as exc:
            raise SSEParseError(
                f"Invalid JSON in SSE data field: {exc}",
                raw_data=merged_data,
            ) from exc

        # Silently drop ping events -- they are transport-layer keepalives
        if event_type == "ping" or parsed_data.get("type") == "ping":
            return None

        return ParsedEvent(event=event_type, data=parsed_data)
